### ISSUE TITLE
Cognito confirmForgotPassword AWS PHP SDK not working

### ISSUE THREAD

**Issue by adampblack** [aws/aws-sdk-php#2272]
I have a form which resets the password.
`
$arrForgotPasswordDetails = [
        'ClientId' => getenv('AWS_COGNITO_CLIENT_APP_ID'),
	'Username' => $this->username,
];

$arrResponse = $objAwsClient->forgotPassword($arrForgotPasswordDetails);
`

The code is sends me says "Invalid code provided, please request a code again." when I run below.

`
$arrConfirmForgotPasswordDetails = [
				'ClientId' => getenv('AWS_COGNITO_CLIENT_APP_ID'),
				'ConfirmationCode' => $this->confirmationcode,
				'Password' => $this->password,
				'Username' => $this->username,
			];

			$arrResponse = $objAwsClient->confirmForgotPassword($arrConfirmForgotPasswordDetails);
`

Even when I run the resend confirmation code I don't receive a new text message.

`
$arrResendConfirmationCodeDetails = [
				'ClientId' => getenv('AWS_COGNITO_CLIENT_APP_ID'),
				'Username' => $this->username,
			];

			$arrResponse = $objAwsClient->resendConfirmationCode($arrResendConfirmationCodeDetails);
`

It works fine via the CLI:
cognito-idp forgot-password
ognito-idp confirm-forgot-password

If I reset via the CLI and confirm via PHP it does not work.

If I reset via PHP and confirm via CLI it works.

So my guess is there is a bug within cognito workflow (regarding not send confirmation code) and PHP SDK confirmForgotPassword method.

**Comment by SamRemis** [comment]
Hi there @adampblack,
Thanks for opening an issue- I was able to call these methods without any problem.  I’m not an expert on cognito specifically, so please bear with me.
When you call
`
$arrResponse = $objAwsClient->forgotPassword($arrForgotPasswordDetails);
`
You should never get the following response:
`invalid code provided, please request a code again`
forgotPassword doesn’t actually take a code as input; are you getting that from confirmForgotPassword?

As far as resend confirmation code, I am a bit confused about what you’re trying to do; are you confirming the user for the first time, or trying to resend the forgot password confirmation code?  From [the docs for resend confirmation code](https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-cognito-idp-2016-04-18.html#resendconfirmationcode): “Resends the confirmation (for confirmation of registration) to a specific user in the user pool.”

My best guess is that you're using a stale code for confirmForgotPassword.  Each time you call forgotPassword it sends a new code and invalidates the old.  If there’s a lag on receiving the text message, is it possible that you may have input an older one after requesting a new one?

**Comment by adampblack** [comment]
Thanks for the reply @SamRemis .

To clarify I use forgetPassword no problem and get the text message on my phone.

But when I go to enter it in the form for confirmForgotPassword it says it is invalid confirmation code.

I resolved the issue three fold:
1. I went to SNS text messages - I set the messages to **transactional** as many text messages were failing to be delivered.
2. I increased the **default** SMS cost limit from **$1USD** to $200USD - this required a support increase limit request.
3. I found the cognito resendConfirmationCode only worked for the confirmSignUp not for the confirmForgotPassword - instead I just redid the forgotPassword.
4. For the confirmForgotPassword I ended up using the AWS CLI which seemed more reliable than the PHP SDK.

Rereading your comment - yes I think the confirmation code is only valid for user registration.

The workaround if anyone is having issues with confirmForgotPassword via the PHP SDK is to use the AWS CLI instead.
